File: src/cli_client/cli_client.py
import json
import httpx
import asyncio
from fastapi import FastAPI, HTTPException, Response, status

import re
import logging

from typing import List, Dict, Any

from managers import ToolManualManager
from model import ClientLanguageModel, LanguageModelConfig
from memory_engine import MemoryEngine, MemoryConfig

LOG_FORMAT = "\033[32m%(levelname)s\033[0m:    %(message)s"
logging.basicConfig(level=logging.INFO, format=LOG_FORMAT)
logger = logging.getLogger("cli_client")

ORCHESTRATOR_URL = "http://127.0.0.1:8000/orchestrate"
MEMORY_DIR = "./memory_docs"

# def clean_json_string(raw_string):
#     """Cleans Markdown artifacts and ensures valid JSON (Object or Array) is parsed."""
#     # 1. Remove markdown code block markers
#     cleaned = re.sub(r"```[a-z]*\s?|```", "", raw_string).strip()
    
#     # 2. Find the bounds of the JSON structure (either { } or [ ])
#     # We find the first occurrence of either { or [
#     match_start = re.search(r'[\[\{]', cleaned)
#     # We find the last occurrence of either } or ]
#     match_end = re.search(r'[\}\]](?=[^\}\]]*$)', cleaned)

#     if match_start and match_end:
#         start_idx = match_start.start()
#         end_idx = match_end.start()
#         return cleaned[start_idx:end_idx + 1]
    
#     return cleaned


# TODO: Class Use Agent Sessions to control multiple agent rounds, save memory at the end of memory_session


class CLIClient:
    def __init__(self, model_type="google", model_name=None, test_flag=False):
        self.test_flag = test_flag
        self.tool_manager = ToolManualManager()
        self.language_model = ClientLanguageModel(config=LanguageModelConfig(model_type=model_type, model_name=model_name))
        self.memory_engine = MemoryEngine(config=MemoryConfig(memory_dir=MEMORY_DIR), model_config=self.language_model.config)
        self.memory_buffer_start_id = 0
        self.history_window = 10  # Number of recent turns to keep in direct history context
        
        asyncio.run(self._list_available_servers())
    
    async def _list_available_servers(self):
        async with httpx.AsyncClient() as client:
            response = await client.post(ORCHESTRATOR_URL, json={"id": "auto-listing", "command": "list_available_servers"})
            if response.status_code == status.HTTP_200_OK:
                self.tool_manager.register_summary(json.loads(response["result"]))
                logger.info("Successfully retrieved and registered available servers.")
            else:
                logger.error(f"Error listing servers: {response.status_code} - {response.text}")

    async def agent_loop(self):                
        try:
            await self.memory_engine.aload_memory()
            while True:
                user_input = await asyncio.to_thread(
                    input,
                    "--- User Input ---\nEnter your message for the agent (or 'quit' to exit):\n",
                )
                if user_input.lower() == "quit":
                    await self._flush_memory_buffer()
                    break
                message_id = self.language_model.add_user_message(user_input)

                print(f"\n--- Agent Response ---")
                earliest_message_id = max(0, message_id - self.history_window)
                latest_stored_message_id, recalled_nodes = await self.memory_engine.arecall(
                    query=user_input,
                    earliest_message_id=earliest_message_id
                )
                temp_history_window = message_id - min(earliest_message_id, latest_stored_message_id+1)
                llm_output = self.language_model.parse_response(
                    await self.language_model.agenerate_response(
                        self.tool_manager.get_tool_info(),
                        memory_nodes=recalled_nodes,
                        history_window=temp_history_window
                    )
                )
                if self.test_flag:
                    print(f"LLM Output (with tool calls):\n{llm_output}\n")
                content = llm_output.get("content", "")
                if content != "":
                    print(f"{content}\n")
                tool_calls = llm_output.get("tool_calls", [])
                had_tool_calls = len(tool_calls) > 0
                while len(tool_calls) > 0:
                    await self.tool_callings(tool_calls)
                    temp_history_window = message_id - min(earliest_message_id, latest_stored_message_id+1)
                    llm_output = self.language_model.parse_response(
                        await self.language_model.agenerate_response(
                            self.tool_manager.get_tool_info(),
                            memory_nodes=recalled_nodes,
                            history_window=temp_history_window
                        )
                    )
                    if self.test_flag:
                        print(f"LLM Output (with tool calls):\n{llm_output}\n")
                    content = llm_output.get("content", "")
                    if content != "":
                        print(f"{content}\n")
                    tool_calls = llm_output.get("tool_calls", [])

                # TODO: filter out only important tool results
                # TODO: better summarization triggering algorithm
                if self.memory_buffer_start_id == 0:
                    self.memory_buffer_start_id = message_id
                if self.memory_engine.should_summarize_turn(
                    user_input=user_input,
                    assistant_output=content,
                    had_tool_calls=had_tool_calls,
                    pending_turns=message_id - self.memory_buffer_start_id + 1,
                ):
                    await self._flush_memory_buffer()
        finally:
            await self.memory_engine.asave_memory()

    async def tool_callings(self, tool_calls: List[Dict[str, Any]]):
        logger.info(f"Processing {len(tool_calls)} tool calls...")
        command_records = {}
        calls = []
        # Tool calls preprocess
        for tool_call in tool_calls:
            if "command" not in tool_call or "id" not in tool_call:
                continue
            command_records.update({
                tool_call["id"]: [tool_call["command"], tool_call["arguments"].get("server_name", None)]
            })
            call = tool_call["arguments"]
            call["id"] = tool_call["id"]
            call["command"] = tool_call["command"]
            if call["command"] == "activate_server":
                call["fetch_manual"] = not self.tool_manager.check_tool(call["server_name"])
            calls.append(call)
            logger.info(f"Calling command '{call['command']}' with arguments {call}")

        async with httpx.AsyncClient() as client:
            call_requests = [client.post(ORCHESTRATOR_URL, json=call) for call in calls]
            call_responses = await asyncio.gather(*call_requests)

        # Tool calls reponse postprocess
        for call_resp in call_responses:
            call_resp = json.loads(call_resp.text)
            id = call_resp.get("id")
            command = command_records[id][0]
            server_name = command_records[id][1]
            message_id = 0
            if call_resp["status"] != status.HTTP_200_OK:
                error_status = call_resp["status"]
                detail = call_resp["detail"]
                message_id = self.language_model.add_tool_response(
                    content=f"{error_status} - {detail}",
                    command=command,
                    tool_call_id=id
                )
                logger.error(f"❌ Error in tool call id {id}: {error_status} - {detail}\n")
            else:
                # tool calling results
                if command == "list_available_servers":
                    self.tool_manager.register_summary(json.loads(call_resp["result"]))
                    result = "Tool summaries have been registered."
                if command == "activate_server":
                    if call_resp["result"]:
                        self.tool_manager.register_tool(server_name, call_resp["result"])
                    self.tool_manager.inject_tool(server_name)
                    result = f"Tool '{server_name}' has been activated."
                elif command == "stop_server":
                    self.tool_manager.prune_tool(server_name)
                    result = f"Tool '{server_name}' has been stopped."
                elif command == "execute_server_code":
                    result = call_resp["result"]
                else:
                    result = "Unknown command response."

                message_id = self.language_model.add_tool_response(
                    content=result,
                    command=command,
                    tool_call_id=id
                )
                
                logger.info(f"✅ Success in tool call id {id}")
                logger.info(f"ℹ️  Info: {call_resp.get('info', 'No info available')}\n")
        
        return message_id
    
    # Function: summarize context into memory content
    async def summarize_memory(self, max_memory_tokens: int = 512):
        self.language_model.add_user_message(self._summary_system_prompt(max_memory_tokens))
        response = await self.language_model.agenerate_response(
            self.tool_manager.get_tool_info(),
            memory_nodes=[],
            max_tokens=max_memory_tokens,
        )
        return self.language_model.parse_response(response)

    async def _flush_memory_buffer(self, end_message_id: int):
        if self.memory_buffer_start_id == 0:
            return
        history_window = end_message_id - self.memory_buffer_start_id + 1
        summary = self.memory_engine.summarize_memory_buffer(self.language_model.get_history(window=history_window))
        if summary.strip():
            await self.memory_engine.aremember(context=summary, message_id_range=[self.memory_buffer_start_id, end_message_id])
        self.memory_buffer_start_id = 0


def generate_list_command():
    return """
    ```json
    {
        "command": "list_available_servers"
    }
    ```
    """

def generate_activate_command(server_name):
    return f"""
    ```json
    {{
        "command": "activate_server",
        "server_name": "{server_name}"
    }}
    ```
    """

def generate_execute_commands(server_names, languages, codes):
    calls = []
    for i in range(len(server_names)):
        calls.append({
            "command": "execute_server_code",
            "server_name": server_names[i],
            "language": languages[i],
            "code": codes[i]
        })
    return json.dumps(calls)

def generate_stop_command(server_name):
    return f"""
    ```json
    {{
        "command": "stop_server",
        "server_name": "{server_name}"
    }}
    ```
    """

# Example usage
if __name__ == "__main__":
    cli_client = CLIClient(model_type="openai", test_flag=True)
    asyncio.run(cli_client.agent_loop())


    # cli_client = CLIClient()

    # # Simulated LLM output
    # commands = [
    #     generate_list_command(),
    #     generate_activate_command("weather"),
    #     generate_execute_commands(["weather", "weather"], ["python", "python"], [
    #         "print('Hello from the AI Agent!')",
    #         "import asyncio; print(asyncio.run(get_forecast(34.0522, -118.2437)))"
    #     ]),
    #     generate_stop_command("weather"),
    #     generate_activate_command("weather"),
    #     generate_execute_commands(["weather"], ["python"], [
    #         "import asyncio; print(asyncio.run(get_alerts('MA')))"
    #     ])
    # ]
    
    # for command in commands:
    #     asyncio.run(cli_client.mock_agent_loop(command))

    
    # asyncio.run(cli_client.mock_agent_loop(generate_stop_command("weather")))
    
    

File: src_ts/cli_client/cli_client.ts
import readline from "node:readline";
import process from "node:process";
import { ToolManualManager } from "./managers";
import { LanguageModel, ParsedResponse } from "./model";
import { logError, logInfo, logWarn } from "../shared/logging";
import { CoAdaptSession } from "../coadapt";

const SERVER_URL = "http://127.0.0.1:8000/orchestrate";

type ToolCall = { id: string; command: string; arguments: Record<string, any> };

async function postJson(url: string, body: any): Promise<any> {
  const res = await fetch(url, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(body),
  });
  const text = await res.text();
  return text ? JSON.parse(text) : {};
}

export class CLIClient {
  private tool_manager = new ToolManualManager();
  private language_model: LanguageModel;
  private test_flag: boolean;
  private co_adapt: CoAdaptSession | null;

  constructor(opts?: { model_type?: string; model_name?: string; test_flag?: boolean; co_adapt?: boolean }) {
    const model_type = opts?.model_type ?? "google";
    const model_name = opts?.model_name;
    this.test_flag = !!opts?.test_flag;
    this.language_model = new LanguageModel(model_type, model_name);
    const enable =
      opts?.co_adapt !== false &&
      process.env.COADAPT !== "0" &&
      process.env.COADAPT_DISABLED !== "1";
    this.co_adapt = enable ? new CoAdaptSession() : null;
  }

  async agent_loop(): Promise<void> {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    const question = (q: string) => new Promise<string>((resolve) => rl.question(q, resolve));

    while (true) {
      const user_input = await question("--- User Input ---\nEnter your message for the agent (or 'quit' to exit):\n");
      if (user_input.trim().toLowerCase() === "quit") break;

      if (this.co_adapt) await this.co_adapt.onUserTurnStart(user_input);
      this.language_model.add_user_message(user_input);
      process.stdout.write("\n--- Agent Response ---\n");

      const setCoAdapt = async () => {
        if (!this.co_adapt) {
          this.language_model.set_co_adapt_context("");
          return;
        }
        this.language_model.set_co_adapt_context(await this.co_adapt.buildContextBlock(user_input));
      };
      await setCoAdapt();

      let turn_assistant = "";
      let tool_calls_total = 0;
      let llm_output: ParsedResponse = this.language_model.parse_response(
        await this.language_model.generate_response(this.tool_manager.getToolInfo())
      );

      tool_calls_total += ((llm_output.tool_calls as ToolCall[]) ?? []).length;

      if (this.test_flag) process.stdout.write(`LLM Output (with tool calls):\n${JSON.stringify(llm_output, null, 2)}\n\n`);

      if (llm_output.content) {
        process.stdout.write(`${llm_output.content}\n\n`);
        turn_assistant += `${llm_output.content}\n`;
      }

      let tool_calls: ToolCall[] = (llm_output.tool_calls as any) ?? [];
      while (tool_calls.length > 0) {
        await this.tool_callings(tool_calls);
        await setCoAdapt();
        llm_output = this.language_model.parse_response(
          await this.language_model.generate_response(this.tool_manager.getToolInfo())
        );
        tool_calls_total += ((llm_output.tool_calls as ToolCall[]) ?? []).length;
        if (this.test_flag) process.stdout.write(`LLM Output (with tool calls):\n${JSON.stringify(llm_output, null, 2)}\n\n`);
        if (llm_output.content) {
          process.stdout.write(`${llm_output.content}\n\n`);
          turn_assistant += `${llm_output.content}\n`;
        }
        tool_calls = (llm_output.tool_calls as any) ?? [];
      }

      if (this.co_adapt)
        await this.co_adapt.afterTurn(user_input, turn_assistant.trim(), { toolCallsCount: tool_calls_total });
    }

    rl.close();
  }

  async tool_callings(tool_calls: ToolCall[]) {
    logInfo("cli_client", `Processing ${tool_calls.length} tool calls...`);

    const command_records: Record<string, [string, string | null]> = {};
    const calls: any[] = [];

    for (const tool_call of tool_calls) {
      if (!tool_call.command || !tool_call.id) continue;
      const serverName = tool_call.arguments?.server_name ?? null;
      command_records[tool_call.id] = [tool_call.command, serverName];

      const call: any = { ...(tool_call.arguments ?? {}) };
      call.id = tool_call.id;
      call.command = tool_call.command;
      if (call.command === "activate_server") {
        call.fetch_manual = !this.tool_manager.checkTool(String(call.server_name ?? ""));
      }
      calls.push(call);
      logInfo("cli_client", `Calling command '${call.command}' with arguments ${JSON.stringify(call)}`);
    }

    const responses = await Promise.all(calls.map((c) => postJson(SERVER_URL, c)));

    for (const call_resp of responses) {
      const id = call_resp.id as string;
      const [command, server_name] = command_records[id] ?? ["", null];

      if (call_resp.status !== 200) {
        const error_status = call_resp.status;
        const detail = call_resp.detail;
        this.language_model.add_tool_response(`${error_status} - ${detail}`, command, id);
        logError("cli_client", `Error in tool call id ${id}: ${error_status} - ${detail}`);
        continue;
      }

      let resultForModel = "Unknown command response.";
      if (command === "list_available_servers") {
        try {
          this.tool_manager.registerSummary(JSON.parse(call_resp.result));
        } catch {
          // ignore
        }
        resultForModel = "Tool summaries have been registered.";
      } else if (command === "activate_server") {
        if (call_resp.result && server_name) this.tool_manager.registerTool(server_name, call_resp.result);
        if (server_name) this.tool_manager.injectTool(server_name);
        resultForModel = `Tool '${server_name}' has been activated.`;
      } else if (command === "stop_server") {
        if (server_name) this.tool_manager.pruneTool(server_name);
        resultForModel = `Tool '${server_name}' has been stopped.`;
      } else if (command === "execute_server_code") {
        resultForModel = String(call_resp.result ?? "");
      }

      this.language_model.add_tool_response(resultForModel, command, id);
      logInfo("cli_client", `Success in tool call id ${id}`);
      if (call_resp.info) logWarn("cli_client", `Info: ${call_resp.info}`);
    }
  }
}

if (require.main === module) {
  const client = new CLIClient({ model_type: process.env.MODEL_TYPE ?? "openai", test_flag: true });
  client.agent_loop().catch((e) => {
    logError("cli_client", e?.message ?? String(e));
    process.exitCode = 1;
  });
}

